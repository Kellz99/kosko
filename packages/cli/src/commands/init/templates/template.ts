import fs from "fs";
import { join } from "path";
import { File } from "./base";

const TEMPLATE_DIR = join(__dirname, "../../../../templates");

export async function generateFromTemplateFile(
  path: string,
  template = path
): Promise<File> {
  return {
    path,
    content: await fs.promises.readFile(join(TEMPLATE_DIR, template), "utf8")
  };
}

export function generateReadme() {
  return generateFromTemplateFile("README.md");
}
