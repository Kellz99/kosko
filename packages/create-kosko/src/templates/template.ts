import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { File } from "./base";
import { BUILD_PROD } from "@kosko/build-scripts";

const TEMPLATE_DIR = join(
  fileURLToPath(import.meta.url),
  BUILD_PROD ? "../../templates" : "../../../templates"
);

export async function generateFromTemplateFile(
  path: string,
  template = path
): Promise<File> {
  return {
    path,
    content: await readFile(join(TEMPLATE_DIR, template), "utf8")
  };
}

export function generateReadme() {
  return generateFromTemplateFile("README.md");
}
