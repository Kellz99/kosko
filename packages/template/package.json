{
  "name": "@kosko/template",
  "version": "2.0.1",
  "description": "Utilities for kosko templates.",
  "homepage": "https://kosko.dev",
  "repository": "https://github.com/tommy351/kosko/tree/master/packages/template",
  "author": "Tommy Chen <tommy351@gmail.com>",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    "import": "./dist/index.mjs",
    "require": "./dist/index.js"
  },
  "files": [
    "/dist/"
  ],
  "keywords": [
    "kosko"
  ],
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": ">=14"
  },
  "dependencies": {
    "@kosko/log": "workspace:^0.1.0",
    "exit": "^0.1.2",
    "tslib": "^2.4.0",
    "yargs": "^17.5.1"
  },
  "devDependencies": {
    "@kosko/test-utils": "workspace:^0.0.0",
    "@kosko/jest-preset": "workspace:^0.0.0",
    "@types/exit": "^0.1.31",
    "@types/yargs": "^17.0.12",
    "execa": "^5.1.1"
  },
  "jest": {
    "preset": "@kosko/jest-preset"
  }
}
