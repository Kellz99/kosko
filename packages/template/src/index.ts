export * from "./cli";
export * from "./template";
export * from "./write";
