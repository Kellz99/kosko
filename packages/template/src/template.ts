import yargs from "yargs";

export interface Template<T> {
  description?: string;
  options?: { [P in keyof T]: Options<T[P], T> };
  generate(options: T): Promise<Result>;
}

type OptionKeyRefs<T> =
  | T
  | ReadonlyArray<T>
  | { [key: string]: T | ReadonlyArray<T> };

type Options<T, O> = yargs.Options & {
  coerce?: (arg: any) => T;
  choices?: ReadonlyArray<T>;
  default?: T | (() => T | undefined);
  conflicts?: OptionKeyRefs<keyof O>;
  implies?: OptionKeyRefs<keyof O>;
} & (T extends string
    ? { type: "string" } | { string: true }
    : T extends number
    ? { type: "number" | "count" } | { number: true } | { count: true }
    : T extends boolean
    ? { type: "boolean" } | { boolean: true }
    : T extends any[]
    ? { type: "array" } | { array: true }
    : {});

export interface Result {
  files: ReadonlyArray<File>;
}

export interface File {
  /**
   * Destination path of a file.
   */
  path: string;

  /**
   * File content.
   */
  content: string;
}
