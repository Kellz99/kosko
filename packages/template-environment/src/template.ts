import { Template } from "@kosko/template";
import { join } from "node:path";
import { BUILD_TARGET } from "@kosko/build-scripts";

/**
 * @public
 */
export interface Options {
  name: string;
  esm: boolean;
}

const esmContent = `export default {}`;

const cjsContent = `"use strict";

module.exports = {};
`;

/**
 * @public
 */
export const template: Template<Options> = {
  description: "Create a new environment",
  options: {
    name: {
      type: "string",
      description: "Environment name",
      required: true
    },
    esm: {
      type: "boolean",
      description: "Generate ECMAScript module (ESM) files",
      default: BUILD_TARGET !== "node"
    }
  },
  async generate({ name, esm }) {
    return {
      files: [
        {
          path: join("environments", name, "index.js"),
          content: esm ? esmContent : cjsContent
        }
      ]
    };
  }
};
