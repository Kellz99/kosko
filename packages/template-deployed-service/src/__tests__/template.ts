import { template } from "../template";

test("should return files", async () => {
  const result = await template.generate({
    name: "foo",
    image: "foo:bar",
    type: "NodePort",
    servicePort: 1234,
    containerPort: 5678,
    replicas: 3
  });

  expect(
    result.files.map((file) => ({
      ...file,
      path: file.path.replace(/\\/g, "/")
    }))
  ).toMatchSnapshot();
});
