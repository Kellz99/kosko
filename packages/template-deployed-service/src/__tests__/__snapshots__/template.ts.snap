// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return files 1`] = `
{
  "files": [
    {
      "content": ""use strict";

const { Deployment } = require("kubernetes-models/apps/v1/Deployment");
const { Service } = require("kubernetes-models/v1/Service");

const metadata = { name: "foo" };
const labels = { app: "foo" };

const deployment = new Deployment({
  metadata,
  spec: {
    replicas: 3,
    selector: {
      matchLabels: labels
    },
    template: {
      metadata: {
        labels
      },
      spec: {
        containers: [
          {
            image: "foo:bar",
            name: "foo",
            ports: [{
              containerPort: 5678
            }]
          }
        ]
      }
    }
  }
});

const service = new Service({
  metadata,
  spec: {
    selector: labels,
    type: "NodePort",
    ports: [
      {
        port: 1234,
        targetPort: 5678
      }
    ]
  }
});

module.exports = [deployment, service];
",
      "path": "components/foo.js",
    },
  ],
}
`;
